vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,259 @@
+        int x0, y0, x1, y1;  // extent of the region
+        int n_blks, n_tiles; // number of blocks and compatible tiles
+        bool overused(float beta) const {
+            // determines whether region is overutilized:
+            // overused = (Occupancy / Capacity) > beta
+            if (n_blks > beta * n_tiles) return true;
+            return false;
+        }
+    };
+
+    // Utilization of each tile, indexed by x, y
+    std::vector<std::vector<int>> occupancy;
+
+    // Region ID of each tile, indexed by x, y. -1 if not covered by any region
+    std::vector<std::vector<int>> groups;
+
+    // Extent of macro at x, y location. If blk is not in any macros, MacroExtent only covers a single locations
+    std::vector<std::vector<MacroExtent>> macros;
+
+    // List of logic blocks of blk_type at x, y location, indexed by x, y
+    std::vector<std::vector<std::vector<ClusterBlockId>>> blks_at_location;
+
+    // List of sub_tiles compatible with blk_type at x, y location, indexed by x, y
+    std::vector<std::vector<std::vector<t_pl_loc>>> ft;
+
+    // List of all SpreaderRegion, index of vector members is the id of the region
+    std::vector<SpreaderRegion> regions;
+
+    // List of all merged_regions, these regions are merged in larger regions and should be skipped when
+    // recursively cur_spreading. Each entry is the region's ID, which is also the index into the regions vector
+    std::unordered_set<int> merged_regions;
+
+    // Lookup of macro's extent by block ID. If block is a single block, MacroExtent contains only 1 tile location
+    std::map<ClusterBlockId, MacroExtent> blk_extents;
+
+    // Setup CutSpreader data structures using information from AnalyticPlacer
+    // including blks_at_location, macros, groups, etc.
+    void init();
+
+    // Returns number of logical blocks at x, y location
+    int occ_at(int x, int y);
+
+    // Returns number of compatible sub_tiles at x, y location
+    int tiles_at(int x, int y);
+
+    /*
+     * Merge mergee into merged by:
+     * * change group id at mergee grids to merged id
+     * * adds all n_blks and n_tiles from mergee to merged region
+     * * grow merged to include all mergee grids
+     */
+    void merge_regions(SpreaderRegion& merged, SpreaderRegion& mergee);
+
+    /*
+     * Grow region r to include a rectangular region (x0, y0, x1, y1)
+     * Pass init = true if first time calling for a newly created region
+     */
+    void grow_region(SpreaderRegion& r, int x0, int y0, int x1, int y1, bool init = false);
+
+    /*
+     * Expand all utilized regions until they satisfy n_tiles * beta >= n_blocks
+     * If overutilized regions overlap in this process, they are merged
+     */
+    void expand_regions();
+
+    /*
+     * Find overutilized regions surrounded by non-overutilized regions
+     * Start off at an overutilized tile and expand in x, y directions until the region is surrounded by
```

Comment:
Do you expand alternately in the x and y directions (or equally in both directions)?  Good to explain briefly how the x,y growth works.